### ISSUE TITLE
compile problems on Android ndk r16 beta 1

### ISSUE THREAD

**Issue by berkus** [boostorg/filesystem#56]
Hello, in attempt to build boost for android with more or less reasonable strict settings (-Werror) I uncovered some bugs.

```
/usr/local/opt/android-ndk/android-ndk-r16-beta1//sources/android/support/include/stdio.h:36:25: error: invalid token at start of a preprocessor expression
#if __USE_FILE_OFFSET64 && __ANDROID_API__ < __ANDROID_API_N__
```

This happens because filesystem defines a "harmless" macro:

```
libs/filesystem/src/operations.cpp:18:9: warning: '__USE_FILE_OFFSET64' macro redefined [-Wmacro-redefined]
#define __USE_FILE_OFFSET64 // but that is harmless on Windows and on POSIX
        ^
```

While this check was `#if defined(__USE_FILE_OFFSET64)` in previous android versions, it seems to have changed when they introduced unified headers. It's probably more "harmless" to define an integer value in boost.filesystem rather than persuade Google to fix their code.

This is using `clang` toolchain on `Android NDK r16 beta1`.

[​Reference to user-config.jam for the build](https://github.com/twilio/twilio-boost-build/blob/android-ndk-16/boost.sh#L526)

**Comment by berkus** [comment]
Xref https://github.com/boostorg/filesystem/issues/56

**Comment by pdimov** [comment]
See also https://github.com/android-ndk/ndk/issues/518

**Comment by enh** [comment]
this bug can be closed: the released NDK added the missing `defined()` last year.

**Comment by alexeikh** [comment]
I should be noted that:
- This issue (#56) is about compilation failure of `#include <cstdio>` if `_FILE_OFFSET_BITS=64` is defined.
- This issue (#56) should not be confused with issue #65, which is about compilation failure of `BOOST_RESIZE_FILE(...)` if `_FILE_OFFSET_BITS=64` is defined.
- This issue (#56) is only fixed in Android NDK for libc++ (LLVM STL). If using libstdc++ (GNU STL), the issue remains.
- Issue #65 is not fixed yet. Android NDK developers' opinion is that it must be fixed in the Boost.Filesystem code.
- Pull request #69 aims to fix both issue #56 for GNU STL and issue #65.

I am publishing this information here, so that #65 and #69 will not be mistakenly closed as duplicates for this issue (#56).

**Comment by berkus** [comment]
@alexeikh libstdc++ is no longer supported in latest NDKs and is slated for removal. I don't think any significant work will be done to remedy this in libstdc++.
